### PR TITLE
Format Datasets into DCAT Feed

### PR DESCRIPTION
**Pull request by andrewctate**
This PR ports the hub-indexer [code](https://github.com/ArcGIS/hub-indexer/blob/c33ac0b73b44e7efb92f551339ffbed3d9897a16/packages/api/responders/helpers/adlib-dcat.js) for formatting datasets to our Koop output plugin. Custom DCAT configurations are not supported but will be added in a separate PR.

### CODE REVIEW

File: src/dcat-us/dataset-formatter.ts [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+interface DistributionParameters {
+  landingPage: string;
+  serviceUrl: string;
+  downloadLinkFor: (type: string) => string;
+}
+
+/*
+* Generate DCAT Distributions
+*/
+function generateDistributions (dataset: any, landingPage: string) {
```
Comment (andrewctate) [inline src/dcat-us/dataset-formatter.ts:92]: This file got long. We could split all the distribution code out into its own place if that's better.

File: src/dcat-us/base-template.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+// the base DCAT adlib template
+export const baseTemplate = {
```
Comment (andrewctate) [inline src/dcat-us/base-template.ts:2]: Rename to be `baseDatasetTemplate`

File: src/dcat-us/dataset-formatter.ts [hunk 3]
```diff
@@ -0,0 +1,250 @@
+import * as _ from 'lodash';
+import { adlib, TransformsList } from 'adlib';
+import { isPage } from '@esri/hub-sites';
+import { baseTemplate } from './base-template';
+
+export function formatDcatDataset (dataset: any, siteUrl: string, dcatCustomizations: any = {}) {
```
Comment (andrewctate) [inline src/dcat-us/dataset-formatter.ts:6]: rename to `hubDataset`

File: src/dcat-us/dataset-formatter.ts (hunk 1 above)
Comment (sonofflynn89) [reply src/dcat-us/dataset-formatter.ts:92]: I agree. I think the distribution logic should be broken out into its own file with its own tests for future developer clarity.

**Review by sonofflynn89 — CHANGES_REQUESTED** [review]
Just move the distro logic to another file and rename 'dataset' to 'hubDataset' and everything will be good! Thanks for doing this!

**Review by rgwozdz — COMMENTED** [review]
See comments.  

Lots of use of `any` as a type.  Can we use more specific types?

File: src/dcat-us/dataset-formatter.ts [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    || !dcatDataset.keyword[0]; // if tags is undefined, the tags array is empty, or tags is an empty string
+
+  if (isPage(dataset) && hasNoTags) {
+    dcatDataset.keyword = ['ArcGIS Hub page'];
+  }
+
+  dcatDataset.distribution = generateDistributions(dataset, landingPage);
+
+  // only add spatial property if there's an extent
+  if (dataset.extent && dataset.extent.coordinates) {
```
[the remark this answers is not in the document]
Comment (andrewctate) [reply src/dcat-us/dataset-formatter.ts:44]: Just a copy-paste. I'll update all the same.

File: src/dcat-us/dataset-formatter.ts [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+}
+
+// HUBJS CANDIDATE
+function indent(str: string, nTabs = 1) {
+  const tabs = new Array(nTabs).fill('\t').join('');
+  return tabs + str.replace(/\n/g, `\n${tabs}`);
+}
+
+// remove customizations that attempt to overwrite special or required fields
+function scrub (customizations: any) {
```
[the remark this answers is not in the document]
Comment (andrewctate) [reply src/dcat-us/dataset-formatter.ts:64]: Another copy-paste. We could scope the type down to `Record<any, string>`

File: src/dcat-us/dataset-formatter.ts [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    delete customizations.identifier;
+    delete customizations.landingPage;
+    delete customizations.webService;
+    if (customizations.contactPoint) {
+      delete customizations.contactPoint['@type'];
+    }
+    delete customizations.distributions;
+  }
+
+  return customizations;
```
[the remark this answers is not in the document]
Comment (andrewctate) [reply src/dcat-us/dataset-formatter.ts:76]: Copy-paste, but I'll look into this!

File: src/dcat-us/dataset-formatter.ts [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+    downloadLinkFor: getDownloadLinkFn(landingPage, dataset)
+  };
+
+  const distributions = distributionFns.map(fn => fn(params));
+  const customDistributions = getCustomDistributions(dataset);
+
+  return [...distributions, ...customDistributions];
+}
+
+function isLayer (dataset: any) {
```
[the remark this answers is not in the document]
Comment (andrewctate) [reply src/dcat-us/dataset-formatter.ts]: No one has defined a type for the JSON API v3 response. Felt like it was outside the scope of this story, but I could give it a shot.

**Review by sonofflynn89 — APPROVED** [review]
Thanks for moving the distributions logic! It looks great!

**Pull request merged**